feat(kad)!: add `ModeChanged` event

dhuseby opened the pull request:
## Description

Previously, the kademlia protocol would only log changes to the internal mode. With this patch, we now also emit an event which allows users to code against the internal state of the kademlia protocol.

Fixes #4310.

## Notes & open questions

I added the code to emit the event in the reconfigure_mode function on the assumption that all mode changes result in that function being called.

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

dhuseby committed:
feat!: added kademlia ModeChanged event

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

dhuseby committed:
feat!: added kademlia ModeChanged event

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

dhuseby committed:
bumped crate version

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

jxs committed:
deps(identity): update ed25519-dalek to 2.0

Pull-Request: #4337.

thomaseizinger commented on protocols/kad/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.45.0
```
```suggestion
## 0.45.0 - unreleased
```

thomaseizinger commented on protocols/kad/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.45.0
+
+- Added ModeChanged behaviour event.
```
```suggestion
- Emit `ModeChanged` event whenever we automatically reconfigure the mode.
```

thomaseizinger commented on protocols/kad/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+## 0.45.0
+
+- Added ModeChanged behaviour event.
```
Missing PR reference, see other changelog entries.

dhuseby replied:
added

thomaseizinger commented on protocols/kad/src/behaviour.rs:
```diff
@@ -2674,6 +2679,12 @@ pub enum KademliaEvent {
+    ///
+    /// This usually happens in response to an address change or a new external
+    /// address being added or removed.
+    ModeChanged { mode: Mode },
```
```suggestion
    ModeChanged { new_mode: Mode },
```

thomaseizinger reviewed (commented):
Thanks!

The test now fails which is good. We should update the test to also assert the new event being emitted.

thomaseizinger commented on protocols/kad/src/behaviour.rs:
```diff
@@ -1045,6 +1045,11 @@ where
+        self.queued_events
+            .push_back(ToSwarm::GenerateEvent(KademliaEvent::ModeChanged {
+                mode: self.mode,
+            }));
```
This is also called when we set the mode manually, see line 1014. I think we should move this to `determine_mode_from_external_addresses`.

thomaseizinger replied:
We'll need to slightly modify that function though because at the moment, we don't know whether the mode actually changed at the end. We'll need to perform some diffing there and only emit the event if we actually switched from client to server or vice-versa.

dhuseby replied:
done

mxinden committed:
chore: use workspace `rust-version`

Pull-Request: #4344.

p-shahi committed:
docs: remove archived `ursa` project from README

Pull-Request: #4340.

jxs committed:
refactor(tcp): reducing branching in `Transport::create_socket`

Following https://github.com/libp2p/rust-libp2p/pull/4289#discussion_r1285812762, hereby is the PR to also improve the `create_socket`  using [`for_addr`](https://docs.rs/socket2/latest/socket2/struct.Domain.html#method.for_address). We also add a test for listening on IPv4 and IPv6 separately.

Pull-Request: #4328.

yerayrodriguez committed:
docs(readme): add taple to notable users

Request to add [TAPLE](https://www.taple.es) Project to rust-libp2p notable users. TAPLE is a permissioned DLT solution for traceability of assets and processes. We use libp2p for the network layer.

Pull-Request: #4309.

dependabot[bot] committed:
deps: bump pin-project from 1.1.2 to 1.1.3

Pull-Request: #4296.

dhuseby committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dhuseby committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dhuseby committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dhuseby committed:
Update protocols/kad/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dhuseby committed:
Update protocols/kad/src/behaviour.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dhuseby committed:
Update protocols/kad/src/behaviour.rs

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dhuseby committed:
event only emits when mode changes automatically

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

dhuseby committed:
rebasing

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

dhuseby committed:
catching up with GH suggested changes

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

dhuseby committed:
pulling in GH merges messed things up a bit

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

dhuseby committed:
fix the broken tests

Signed-off-by: Dave Huseby <dwh@linuxprogrammer.org>

thomaseizinger commented on README.md:
```diff
@@ -103,4 +103,4 @@ used and maintained by [Actyx](https://www.actyx.com).
 - [Substrate](https://github.com/paritytech/substrate) - Framework for blockchain innovation,
 used by [Polkadot](https://www.parity.io/technologies/polkadot/).
-- [Ursa](https://github.com/fleek-network/ursa) - Decentralized Content Delivery & Edge Compute.
+- [Taple](https://github.com/opencanarias/taple-core) - Sustainable DLT for asset and process traceability by [OpenCanarias](https://www.opencanarias.com/en/).
```
Something seems to have gone wrong during the rebase / merge because this diff is showing up here as if it were in this PR?

dhuseby replied:
yeah...I think I pulled the upstream master into my branch at some point. let me see if I can clean this up.

dhuseby replied:
I tried to do an interactive rebase and drop the unwanted commits and they don't show up in the list of available commits to drop. So instead I created a new branch and used cherry-pick to recreate this PR's branch with just my commits and your review commits. The new branch is here: https://github.com/dhuseby/rust-libp2p/tree/add-kad-modechanged-event

I was think the best course of action at this point is to close this PR and file a new PR with the new clean branch. What do you think?

thomaseizinger replied:
> I was think the best course of action at this point is to close this PR and file a new PR with the new clean branch. What do you think?

Yes, sounds good!
